File: compile.rb
#!/usr/bin/ruby -w

$PRO = "cgraph"
$DIR = "."
$SRC = File.join($DIR, "src")
$INC = File.join($DIR, "include")
$TST = File.join($DIR, "test")
$LIB = File.join($DIR, "lib")

$CC = "cc"
$CFLAGS = "-pedantic -Wall -fpic -std=c89"
$CSFLAGS = "-shared"

$MODE = "debug"
if $MODE == "debug" then
  $CFLAGS = "$CFLAGS -g -DDEBUG"
elsif $MODE == "release" then
  $CFLAGS = "$CFLAGS -static -O2"
end

# package shared library
$AR = "ar"
$ARFLAGS = "-rcs"

$CFILES = Dir["*.c"]
print()

$args = ARGV
if $args.size == 0
  if not File.directory?($LIB)
    Dir::mkdir($LIB)
  end
  for $file in $CFILES do
    $obj = $file.sub!(".c$", ".o")
    puts("compile %s to %s", File.join($SRC, $file), File.join($SRC, $obj))
    system(puts("$CC $CFLAGS -I$INC -c %s -o %s", File.join($SRC, $file), File.join($SRC, $obj)))
  end
  puts("compile $LIBSHARED")
  system(sprintf("$CC $CSFLAGS -o $LIBSHARED %s/*.o", $SRC))
  puts("compile $LIBSTATIC")
  system(sprintf("$AR $ARFLAGS $LIBSTATIC %s/*.o", $SRC))
elsif $args[0] == "test"
  puts("compile $TSTFILE to $TSTTARGET")
  system("$CC $CFLAGS -I$INC -o $TSTTARGET $TSTFILE -L$LIB -static -l$PRO -lm")
  printf("test $TSTTARGET with %s", File.join($TST, "elements.csv"))
  system(sprintf("$TSTTARGET %s", File.join($TST, "elements.csv")))
elsif $args[0] == "clean"
  for $file in $CFILES do
    $obj = $file.sub!(".c$", ".o")
    puts("clean $obj")
    File::delete(File.join($SRC, $obj))
  end
  puts("clean $LIBSTATIC")
  File::delete($LIBSTATIC)
  puts("clean $LIBSHARED")
  File::delete($LIBSHARED)
  puts("clean $TSTTARGET")
  File::delete($TSTTARGET)
elsif $args[0] == "distclean"
  for $file in $CFILES do
    my $obj = $file.sub!(".c$", ".o")
    puts("clean $obj")
    File::delete(File.join($SRC, $obj))
  end
  puts("clean $LIBSTATIC")
  File::delete($LIBSTATIC)
  puts("clean $LIBSHARED")
  File::delete($LIBSHARED)
  puts("clean $LIB")
  Dir::rmdir($LIB)
  puts("clean $TSTTARGET")
  File::delete($TSTTARGET)
elsif $args[0] == "help"
  puts("__FILE__ <target>")
  puts("<target>: ")
  puts("                    compile cgraph")
  puts("          test      test cgraph\n")
  puts("          clean     clean all the generated files")
  puts("          distclean clean all the generated files and directories")
  puts("          help      commands to this program")
else
  puts("$args[0] is an unsupported command")
  printf("use \"%s help\" to know all supported commands", __FILE__)
end

File: compile.pl
#!/usr/bin/perl -w

use strict;
use warnings;

my $PRO = "cgraph";
my $DIR = ".";
my $SRC = "$DIR/src";
my $INC = "$DIR/include";
my $TST = "$DIR/test";
my $LIB = "$DIR/lib";

my $CC = "cc";
my $CFLAGS = "-pedantic -Wall -fpic -std=c89";
my $CSFLAGS = "-shared";

my $MODE = "debug";
if ($MODE eq "debug")
{
  $CFLAGS = "$CFLAGS -g -DDEBUG";
}
elsif ($MODE eq "release")
{
  $CFLAGS = "$CFLAGS -static -O2";
}

# package shared library
my $AR = "ar";
my $ARFLAGS = "-rcs";

# source files
my @CFILES;
opendir(my $fin, $SRC) or die "cannot open directory $SRC";
foreach my $file (readdir($fin))
{
  if($file =~ /\.c$/)
  {
    push(@CFILES, $file);
  }
}
closedir($fin);
my $LIBSHARED;
my $LIBSTATIC;
my $TSTFILE;
my $TSTTARGET;
if ($^O eq "MSWin32")
{
  # target files
  $LIBSHARED = "$LIB/lib$PRO.dll";
  $LIBSTATIC = "$LIB/lib$PRO.a";
  # test files
  $TSTFILE = "$TST/$PRO.c";
  $TSTTARGET = "$TST/$PRO.exe";
}
else
{
  # target files
  $LIBSHARED = "$LIB/lib$PRO.so";
  $LIBSTATIC = "$LIB/lib$PRO.a";
  # test files
  $TSTFILE = "$TST/$PRO.c";
  $TSTTARGET = "$TST/$PRO";
}

my @args = @ARGV;

if ($args[0] eq "")
{
  mkdir $LIB;
  foreach my $file (@CFILES)
  {
    my $obj = ($file =~ s/\.c/\.o/r);
    print("compile $SRC/$file to $SRC/$obj\n");
    system("$CC $CFLAGS -I$INC -c $SRC/$file -o $SRC/$obj");
  }
  print("compile $LIBSHARED\n");
  system("$CC $CSFLAGS -o $LIBSHARED $SRC/*.o");
  print("compile $LIBSTATIC\n");
  system("$AR $ARFLAGS $LIBSTATIC $SRC/*.o");
}
elsif ($args[0] eq "test")
{
  print("compile $TSTFILE to $TSTTARGET\n");
  system("$CC $CFLAGS -I$INC -o $TSTTARGET $TSTFILE -L$LIB -static -l$PRO -lm");
  print("test $TSTTARGET with $TST/elements.csv\n");
  system("$TSTTARGET $TST/elements.csv");
}
elsif ($args[0] eq "clean")
{
  foreach my $file (@CFILES)
  {
    my $obj = ($file =~ s/\.c/\.o/r);
    print("clean $obj\n");
    unlink "$SRC/$obj";
  }
  print("clean $LIBSTATIC\n");
  unlink "$LIBSTATIC";
  print("clean $LIBSHARED\n");
  unlink "$LIBSHARED";
  print("clean $TSTTARGET\n");
  unlink "$TSTTARGET";
}
elsif ($args[0] eq "distclean")
{
  foreach my $file (@CFILES)
  {
    my $obj = ($file =~ s/\.c/\.o/r);
    print("clean $obj\n");
    unlink "$SRC/$obj";
  }
  print("clean $LIBSTATIC\n");
  unlink "$LIBSTATIC";
  print("clean $LIBSHARED\n");
  unlink "$LIBSHARED";
  print("clean $LIB\n");
  rmdir $LIB;
  print("clean $TSTTARGET\n");
  unlink "$TSTTARGET";
}
elsif ($args[0] eq "help")
{
  print("$0 <target>\n");
  print("<target>: \n");
  print("                    compile cgraph\n");
  print("          test      test cgraph\n");
  print("          clean     clean all the generated files\n");
  print("          distclean clean all the generated files and directories\n");
  print("          help      commands to this program\n");
}
else
{
  print("$args[0] is an unsupported command\n");
  print("use \"$0 help\" to know all supported commands\n");
}